### PR TITLE
Link: port to new ethtool ETHTOOL_xLINKSETTINGS 

### PR DESCRIPTION
**Pull request by ssahani**
This patch defines a new ETHTOOL_GLINKSETTINGS/SLINKSETTINGS API,
handled by the new get_link_ksettings/set_link_ksettings .

This is a WIP version based on this [kernel patch](https://patchwork.kernel.org/patch/8411401/).

### CODE REVIEW

**Comment by martinpitt** [comment]
```
../src/udev/net/ethtool-util.h:56:38: error: field ‘base’ has incomplete type
         struct ethtool_link_settings base;
```

This depends on the linked kernel patch. As this is not even in master (let alone in a two year old kernel which we support in systemd) this needs some `#ifdef`ery to also work with released kernels.

**Comment by ssahani** [comment]
The build is getting passed on fedora  4.7.2-201.fc24.x86_64 . `struct ethtool_link_settings` seems not present in the `/usr/include/linux/ethtool.h`.

**Comment by poettering** [comment]
Let's look into this as soon as the kernel patch has been merged upstream (or has it yet?)

And when that happened, please add the structure to missing.h so that this also compiles on older systems.

File: man/systemd.link.xml [hunk 1]
```diff
@@ -345,9 +345,7 @@
       <varlistentry>
         <term><varname>BitsPerSecond=</varname></term>
         <listitem>
-          <para>The speed to set for the device, the value is rounded
-          down to the nearest Mbps. The usual suffixes K, M, G, are
-          supported and are understood to the base of 1000.</para>
+          <para>The speed to set for the device in Mbps.</para>
```
Comment (poettering) [inline man/systemd.link.xml]: why drop the suffix support?

**Comment by ssahani** [comment]
I think if the structure is not  there we should retain to the old code and if it's there we adopt the new one.

File: man/systemd.link.xml (hunk 1 above)
Comment (ssahani) [reply man/systemd.link.xml]: Not sure But I think other than Mbps is supported. Looking at the man of ethtool

```
speed N
Set speed in Mb/s. ethtool with just the device name as an argument will show you the supported device speeds.

```

File: configure.ac [hunk 2]
```diff
@@ -359,6 +359,12 @@ AC_CHECK_DECLS([IFLA_INET6_ADDR_GEN_MODE,
 #include <linux/neighbour.h>
 ]])
 
+AC_CHECK_TYPE([struct ethtool_link_settings],
+ [AC_DEFINE([HAVE_STRUCT_ETHTOOL_LINK_SETTINGS],[1],
+    [Defined if struct ethtool_link_settings exists in linux/ethtool.h])],
+  [], [[#include <linux/ethtool.h>
+]])
+
```
Comment (keszybz) [inline configure.ac]: It should be enough to add this to AC_CHECK_TYPES above.

File: man/systemd.link.xml (hunk 1 above)
Comment (keszybz) [reply man/systemd.link.xml]: Suffix should continue to be supported. Just convert to the right units internally.

File: man/systemd.link.xml [hunk 3]
```diff
@@ -359,6 +357,17 @@
         </listitem>
       </varlistentry>
       <varlistentry>
+        <term><varname>AutoNegotiation=</varname></term>
+        <listitem>
+          <para>The AutoNegotiation is an ethernet procedure by which two connected devices
+          choose common transmission parameters, such as speed, duplex mode, and flow control.
+          Takes a boolean value. Defaults to "unset".</para>
```
Comment (keszybz) [inline man/systemd.link.xml]: "Unset by defaults, which means that the kernel default will be used."

Please add an empty line between paragraphs.

File: man/systemd.link.xml [hunk 4]
```diff
@@ -359,6 +357,17 @@
...
+          <para> Note that if autonegotiation is disabled, the speed and duplex represent the
```
Comment (keszybz) [inline man/systemd.link.xml]: Spurious space before "Note".

File: src/udev/net/ethtool-util.c [hunk 5]
```diff
@@ -323,3 +265,268 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
 
         return 0;
 }
+
+#ifndef HAVE_STRUCT_ETHTOOL_LINK_SETTINGS
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: This should not be compile time dependent. The code should dynamically fall back to this function if the new code fails. After all, it should be perfectly legal to compile on a newer kernel and run on an older one.

File: src/udev/net/ethtool-util.c [hunk 6]
```diff
@@ -323,3 +265,268 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        };
+        int r;
+
+        ifr->ifr_data = (void *) &ecmd;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        e = malloc0(sizeof(struct ethtool_link_usettings));
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: `new`?

File: src/udev/net/ethtool-util.c [hunk 7]
```diff
@@ -323,3 +265,268 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        return 0;
+}
+
+/* If autonegotiation is disabled, the speed and duplex represent the
+ * fixed link mode and are writable if the driver supports multiple
+ * link modes. If it is enabled then they are read-only; if the link
+ * is up they represent the negotiated link mode; if the link is down,
+ * the speed is 0, %SPEED_UNKNOWN or the highest enabled speed and
+ * @duplex is %DUPLEX_UNKNOWN or the best enabled duplex mode.
+ */
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c:489]: Nowadays we wrap comments to at least 80 columns or so...

File: src/udev/net/ethtool-util.c [hunk 8]
```diff
@@ -323,3 +265,268 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+
+int ethtool_set_glinksettings(int *fd, const char *ifname, unsigned int speed, Duplex duplex, int autoneg) {
+        _cleanup_free_ struct ethtool_link_usettings *u = NULL;
+        struct ifreq ifr = { 0 };
+        int r;
+
+        if (autoneg != 0) {
+                log_info("link_config: autonegotiation is unset or enabled, the speed and duplex are not writable.");
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c:497]: At least ifname should be included in the log message.

File: src/udev/net/ethtool-util.c [hunk 9]
```diff
@@ -323,3 +265,268 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        }
+
+        strscpy(ifr.ifr_name, IFNAMSIZ, ifname);
+
+        r = get_glinksettings(fd, &ifr, &u);
+        if (r < 0) {
+
+                r = get_gset(fd, &ifr, &u);
+                if (r < 0)
+                        return log_warning_errno(r, "link_config: Cannot get device settings: %m");
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: Same here.

File: src/udev/net/ethtool-util.c [hunk 10]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        ifr->ifr_data = (void *) &ecmd;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        if (ecmd.req.link_mode_masks_nwords <= 0 || ecmd.req.cmd != ETHTOOL_GLINKSETTINGS)
+                return -ENOTSUP;
+
+        u = malloc0(sizeof(struct ethtool_link_usettings));
```
Comment (poettering) [inline src/udev/net/ethtool-util.c]: new0()?

File: src/udev/net/ethtool-util.c [hunk 11]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+ * mode and are writable if the driver supports multiple link modes. If it is
+ * enabled then they are read-only; if the link  * is up they represent the
+ * negotiated link mode; if the link is down, the speed is 0, %SPEED_UNKNOWN or
+ * the highest enabled speed and @duplex is %DUPLEX_UNKNOWN or the best enabled
+ * duplex mode.
+ */
+
+int ethtool_set_glinksettings(int *fd, const char *ifname, unsigned int speed, Duplex duplex, int autoneg) {
+        _cleanup_free_ struct ethtool_link_usettings *u = NULL;
+        struct ifreq ifr = { 0 };
```
Comment (poettering) [inline src/udev/net/ethtool-util.c]: `= {}` is sufficient too

File: src/udev/net/link-config.c [hunk 12]
```diff
@@ -380,11 +381,17 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
...
-        r = ethtool_set_speed(&ctx->ethtool_fd, old_name, config->speed / 1024, config->duplex);
-        if (r < 0)
-                log_warning_errno(r, "Could not set speed or duplex of %s to %zu Mbps (%s): %m",
-                                  old_name, config->speed / 1024,
-                                  duplex_to_string(config->duplex));
+        r = ethtool_set_glinksettings(&ctx->ethtool_fd, old_name, config->speed / 1000000, config->duplex, config->autoneg);
+        if (r < 0) {
+
+                if (r == -EOPNOTSUPP)
+                        r = ethtool_set_speed(&ctx->ethtool_fd, old_name, config->speed / 1000000, config->duplex);
```
Comment (poettering) [inline src/udev/net/link-config.c]: i wonder if we shouldn't round up here? i.e. change this to `(config->speed + 999999) / 1000000)`?

(i am just asking this, i really don't know whether rounding up or down is more appropriate here...)

Comment (ssahani) [reply src/udev/net/link-config.c]: I think we can do that `DIV_ROUND_UP(config->speed, 1000000)`

**Comment by ssahani** [comment]
changed and pushed . Hopefully build get passed

File: src/udev/net/ethtool-util.c [hunk 13]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
 
         return 0;
 }
+
+static int get_glinksettings(int *fd, struct ifreq *ifr, struct ethtool_link_usettings **g) {
+        struct ecmd{
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: `ecmd {`

File: src/udev/net/ethtool-util.c [hunk 14]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        return 0;
+}
+
+/* If autonegotiation is disabled, the speed and duplex represent the fixed link
+ * mode and are writable if the driver supports multiple link modes. If it is
+ * enabled then they are read-only; if the link  * is up they represent the
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: `*`

File: src/udev/net/ethtool-util.c [hunk 15]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        ifr->ifr_data = (void *) &ecmd;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        if (ecmd.req.link_mode_masks_nwords >= 0 || ecmd.req.cmd != ETHTOOL_GLINKSETTINGS)
+                return -ENOTSUP;
+
+        ecmd.req.link_mode_masks_nwords = -ecmd.req.link_mode_masks_nwords;
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c:356]: Hm, this ioctl is really strange. Can you add a comment here that explains what is happening and why? Why does the ioctl need to be called twice?

File: src/udev/net/ethtool-util.c [hunk 16]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        ifr->ifr_data = (void *) &ecmd;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        if (ecmd.req.link_mode_masks_nwords <= 0 || ecmd.req.cmd != ETHTOOL_GLINKSETTINGS)
+                return -ENOTSUP;
+
+        u = new0(struct ethtool_link_usettings , sizeof(struct ethtool_link_usettings));
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: Second arg to new0 is the number of elements. It should be 1.

File: src/udev/net/ethtool-util.c [hunk 17]
```diff
@@ -323,3 +324,204 @@ int ethtool_set_features(int *fd, const char *ifname, NetDevFeature *features) {
...
+        };
+        int r;
+
+        ifr->ifr_data = (void *) &ecmd;
+
+        r = ioctl(*fd, SIOCETHTOOL, ifr);
+        if (r < 0)
+                return -errno;
+
+        e = new0(struct ethtool_link_usettings, sizeof(struct ethtool_link_usettings));
```
Comment (keszybz) [inline src/udev/net/ethtool-util.c]: Here too.

File: src/udev/net/link-config.c [hunk 18]
```diff
@@ -167,6 +167,7 @@ static int load_link(link_config_ctx *ctx, const char *filename) {
         link->mac_policy = _MACPOLICY_INVALID;
         link->wol = _WOL_INVALID;
         link->duplex = _DUP_INVALID;
+        link->autoneg = -1;
```
Comment (keszybz) [inline src/udev/net/link-config.c]: autonegotiation, no abbrevs like that please.

File: src/udev/net/link-config.c [hunk 19]
```diff
@@ -380,11 +381,17 @@ int link_config_apply(link_config_ctx *ctx, link_config *config,
         if (!old_name)
                 return -EINVAL;
 
-        r = ethtool_set_speed(&ctx->ethtool_fd, old_name, config->speed / 1024, config->duplex);
-        if (r < 0)
-                log_warning_errno(r, "Could not set speed or duplex of %s to %zu Mbps (%s): %m",
-                                  old_name, config->speed / 1024,
-                                  duplex_to_string(config->duplex));
+        r = ethtool_set_glinksettings(&ctx->ethtool_fd, old_name, DIV_ROUND_UP(config->speed, 1000000), config->duplex, config->autoneg);
```
Comment (keszybz) [inline src/udev/net/link-config.c]: I think you should define a variable for the rounded-up config speed. It seems silly to repeat the same formula three times.

File: src/udev/net/link-config.c (hunk 18 above)
Comment (ssahani) [reply src/udev/net/link-config.c]: ok

File: src/udev/net/link-config.c (hunk 19 above)
Comment (ssahani) [reply src/udev/net/link-config.c]: yes

File: src/udev/net/ethtool-util.c (hunk 15 above)
Comment (ssahani) [reply src/udev/net/ethtool-util.c:356]: This is interesting. User/kernel via the API requires a small ETHTOOL_GLINKSETTINGS  
 handshake

**Comment by poettering** [comment]
So, what's the current state of the kernel patch matching this? Did it get merged/release by now? I figure we shouldn't merge that unless the fate of that patch is clear.

**Comment by keszybz** [comment]
It's upstream: torvalds/linux@3f1ac7a700d039c61d8d8b99f28d605d489a60cf

@ssahani the link to that commit would be nice to include in a comment in the code...

**Comment by ssahani** [comment]
@poettering We have it in fedora already. `4.7.7-200.fc24.x86_64`

@keszybz OK

File: man/systemd.link.xml [hunk 20]
```diff
@@ -359,6 +359,19 @@
         </listitem>
       </varlistentry>
       <varlistentry>
+        <term><varname>AutoNegotiation=</varname></term>
+        <listitem>
+          <para>The AutoNegotiation is an ethernet procedure by which two connected devices
```
Comment (keszybz) [inline man/systemd.link.xml]: `<para>`Enables or disables automatic negotiation of transmission parameters. Autonegotiation is a procedure by which two connected ethernet devices choose common transmission parameters, such as speed, duplex mode, and flow control. Takes a boolean value. Unset by default, which means that the kernel default will be used.`</para>`

`<para>`Note that if autonegotiation is enabled, speed and duplex settings are read-only. If autonegotation is disabled, speed and duplex settings are writable if the driver supports multiple link modes.`</para>`

File: src/udev/net/link-config.c [hunk 21]
```diff
@@ -364,11 +365,12 @@ static int get_mac(struct udev_device *device, bool want_random,
...
                       struct udev_device *device, const char **name) {
-        const char *old_name;
-        const char *new_name = NULL;
+        bool respect_predictable = false;
         struct ether_addr generated_mac;
         struct ether_addr *mac = NULL;
-        bool respect_predictable = false;
+        const char *new_name = NULL;
+        const char *old_name;
+        unsigned int speed;
```
Comment (keszybz) [inline src/udev/net/link-config.c]: Just `unsigned`.

**Pull request merged**